File: solutions/chatbot/java/java-langchain4j/StreamingChatbot.java
/// usr/bin/env jbang "$0" "$@" ; exit $?

//DEPS dev.langchain4j:langchain4j:1.7.1
//DEPS dev.langchain4j:langchain4j-open-ai:1.7.1
//DEPS ch.qos.logback:logback-classic:1.5.6
//FILES ./resources/logback.xml

import java.util.concurrent.CompletableFuture;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import dev.langchain4j.model.chat.StreamingChatModel;
import dev.langchain4j.model.chat.response.ChatResponse;
import dev.langchain4j.model.openai.OpenAiStreamingChatModel;
import dev.langchain4j.service.AiServices;
import dev.langchain4j.service.SystemMessage;
import dev.langchain4j.service.TokenStream;

/**
 * Here is the place where you will add the code to create an advanced chatbot.
 * The steps to create your chatbot are :
 * - choose the right model on AI Endpoints
 * (https://endpoints.ai.cloud.ovh.net), we want to use Mistral 7B Instruct
 * - use the LangChain4J wrapper / lib to do the call to the model (use the AI
 * Service style), ⚠️ don't forget your pom.xml ⚠️
 * - Use the streaming option, see
 * https://docs.langchain4j.dev/tutorials/response-streaming
 * - add parameters to create a virtual assistant named Nestor
 * - ask a question and display the answer in a streaming way
 */

public class StreamingChatbot {
        private static final Logger _LOG = LoggerFactory.getLogger(StreamingChatbot.class);

        // java-06
        // AI Service to create, see https://docs.langchain4j.dev/tutorials/ai-services
        interface Assistant {
                @SystemMessage("You are Nestor, a virtual assistant. Answer to the question.")
                TokenStream chat(String message);
        }

        void main() {
                // java-07
                // Create a streaming chat model using OpenAI provider
                StreamingChatModel steamingModel = OpenAiStreamingChatModel.builder()
                                .apiKey(System.getenv("OVH_AI_ENDPOINTS_ACCESS_TOKEN"))
                                .modelName(System.getenv("OVH_AI_ENDPOINTS_MODEL_NAME"))
                                .baseUrl(System.getenv("OVH_AI_ENDPOINTS_MODEL_URL"))
                                .maxTokens(512)
                                .temperature(0.0)
                                .logRequests(false)
                                .logResponses(false)
                                .build();

                // java-08
                // Build the chatbot thanks to the AIService builder
                // The chatbot must be in streaming mode
                Assistant assistant = AiServices.builder(Assistant.class)
                                .streamingChatModel(steamingModel)
                                .build();

                // java-09
                // Send a prompt
                _LOG.info("💬: Tell me a joke about Java developers\n");
                TokenStream tokenStream = assistant.chat("Tell me a joke about Java developers");
                CompletableFuture<ChatResponse> futureChatResponse = new CompletableFuture<>();

                _LOG.info("🤖: ");
                tokenStream
                                .onCompleteResponse(response -> futureChatResponse.complete(response))
                                .onPartialResponse(_LOG::info)
                                .onError(Throwable::printStackTrace).start();
                futureChatResponse.join();
        }
}

File: solutions/chatbot/python/streaming-chatbot.py
# ⚠️ Don't forget your requirements.txt ⚠️
from langchain_mistralai import ChatMistralAI
import os
from langchain.globals import set_debug

# Activate / deactivate debug messages
set_debug(False)

# py-38
# Initialize the Mistral AI chat model with AI Endpoints, see https://python.langchain.com/docs/integrations/chat/mistralai/
# Use the mistral-7b-instruct-v0-3 model, see https://endpoints.ai.cloud.ovh.net/models/mistral-7b-instruct-v0-3
model = ChatMistralAI(
    base_url=os.getenv("OVH_AI_ENDPOINTS_MODEL_URL"),
    api_key=os.getenv("OVH_AI_ENDPOINTS_ACCESS_TOKEN"),
    model=os.getenv("OVH_AI_ENDPOINTS_MODEL_NAME"),
    max_tokens=512,
    temperature=0
)

# py-39
# Define the messages to send to the model.
# You can set a system prompt and a user message.
messages = [
    (
        "system",
        "You are Nestor, a virtual assistant. Answer to the question.",
    ),
    ("human", "Tell me a joke about Python developers"),
]

# py-40
# Print the response from the model.
print("💬: Question: Tell me a joke about Python developers\n")
print("🤖:")

# py-41
# Call the model with the messages.
# See https://python.langchain.com/docs/how_to/streaming/#llms-and-chat-models
for chunk in model.stream(messages):
    print(chunk.content, end="", flush=True)